### ISSUE TITLE
[MNT] Cut down test time to under 10 min

### ISSUE THREAD

**Issue by lmmentel** [sktime/sktime#2890]
**Describe the bug**

The current testing suite is horrendously long and creates a poor developer experience with the test suite taking ~1h to complete.

**Comment by fkiraly** [comment]
here is an earlier attempt
https://github.com/alan-turing-institute/sktime/pull/2024

but I do not understand why this did not lead to the expected gains. Sth might not work with the caching scope?

**Comment by khrapovs** [comment]
Btw, who is paying for all this computational time on GitHub?.. It's not just about developer experience, is it?

**Comment by fkiraly** [comment]
so, I just notice that in the #2024 attempt, the `lru_cache`-d function is inside a class, which won't work properly.

That might be it?

I.e., the solution would be:

* take #2024 and adapt it into the current test framework, resolving perhaps a million merge conflicts
* take the cached function `_cached_estimator_fitting`, and move it *outside* any of the classes

... done?

**Comment by miraep8** [comment]
I this this is really great. Some ideas to throw out there:

- what would people's thoughts be on just running CI/CD tests on files changed until a particular PR is ready to merge, and then on the merge action we trigger the full CI/CD run? Might help with increasing iteration speed (seems like most bugs come up on files directly related to the PR) but ensures that a full QC check is done before the branch is merged. 
- also with regards to caching and how to store/access the file - it is actually possible to do this via [Github Actions directly](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows).  Not sure if that helps with this particular discussion, but seems like it might.

**Comment by fkiraly** [comment]
>what would people's thoughts be on just running CI/CD tests on files changed until a particular PR is ready to merge, and then on the merge action we trigger the full CI/CD run?

That's a brilliant idea, and basically what `pytest-testmon` does, this avenue was explored in issue https://github.com/alan-turing-institute/sktime/issues/2530.

I think we got stuck at the question "but how would you know which tests are those that touch the changed files/lines"? There is also a problem with `testmon` that it does not detect files that the `all_estimator` crawler does, say, if an entire estimator is added.

I have to apologize to @tarpas who put work into this, we got stuck and I was thinking about how to ensure that really all classes get tested, for weeks, until the point I eventually forgot by accident that the discussion was still ongoing, sorry @tarpas!
(should have tracked that on one of the boards)

**Comment by fkiraly** [comment]
So, @miraep8, do you have a good idea how you would technically implement, given `sktime`'s current test set-up, to test only files changed in a PR?

> also with regards to caching and how to store/access the file - it is actually possible to do this via [Github Actions directly](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows). Not sure if that helps with this particular discussion, but seems like it might.

I also overlooked @tarpas last post which seems to suggest something similar. I.e., first test with the "imperfect" `testmon`, and then gradually add in tests.

The problem though would be that it would miss new files and estimators therein, if I am not mistaken.

But, we could supplement the idea by running `check_estimator` on all classes in new files, then it might cover enough to be practicable?

**Comment by miraep8** [comment]
I think running `check_estimator` on all classes in new files would get us most of the way there.  It might not be ideal for cases where we are modifying `BaseObject` or something like that though.  Thoughts on this suggestion to help us catch all potentially affected files: 

- Lets create an object which is a dict() (lets call it `importsMe`)
- The keys to this object are module names within `sktime`
- The values are a list all the other modules within `sktime` that import a particular module named by the key
- This is something that can easily be created once by crawling through all modules in `sktime` and using `modulefinder`, or `inspect` or something like that to get a list of modules imported.
- I believe the best way to do it might be to list indirect imports as well, eg if `C` imports from `B` which imports from `A`, but `C` does not directly import from `A`, we would still want `C` to show us as a value for key `A`

So my understanding from the current CI/CD is that we currently only run the style checks on files which have been modified.  (I only know this because I thought the current CI/CD is great and copied a lot of the code for it for my own repo!).  But that means we have access to a list of modified modules.  We can now do the following:

- use the changed files to update `importsMe` (ie if any of our modules are new or import from something new we add it to the list of modules importing that particular module).
- lastly we run tests on the changed module, and on all modules which inherit from it. 

Drawback - my current implementation might not handle well if an import gets dropped, but that is something that could be added it if is important. 

Thoughts?

**Comment by fkiraly** [comment]
Summary: I think this is an excellent idea!

We may have to modify it though for functionality that itself uses `import_module`, possibly with dynamic imports.
I think this is isolated to `all_estimators` though, so we may need to think around that.

>Thoughts on this suggestion to help us catch all potentially affected files:

Ok, your `dict` is the transitive closure of the directed acyclic graph of module imports, where directed edges are encoded as (key, value) pairs.

>Drawback - my current implementation might not handle well if an import gets dropped, but that is something that could be added it if is important.

I think that's easy to solve, you build the transitive import graph for the changed and the unchanged version (pre and post PR), and also test everything where an outgoing edge changes.

@tarpas might be able to comment on whether `pytest-testmon` is already doing this or a similar thing.

>I only know this because I thought the current CI/CD is great and copied a lot of the code for it for my own repo!

Where is that? Is that a secret project?

**Comment by tarpas** [comment]
Hi. I prioritized other things because testmon doesn't support pytest --pyargs and also because I'm not able to run the test suite on my M1 Macbook (tensorflow started crashing under rosetta 2 and there are still problems to install it natively)

What @miraep8 describes sounds like static analysis. testmon doesn't do that, it uses only runtime coverage data. That's why it might have problems to pick up constants initialized on module import.

Would you be willing to go away from using --pyargs? That's the show stopper in #2530
As for catching all estimators/classes I assume it would be easy to make the collection a function as opposed to a module constant and that would be the easiest way to allow testmon to catch changes there.

**Comment by fkiraly** [comment]
> it would be easy to make the collection a function

It already is a function! We are using `all_estimator` internally from the registry module.

>Would you be willing to go away from using --pyargs?

Hm, @lmmentel, could you kindly reply to this? I pinged you in the #2530 issue and you did not react, maybe you did not see it?

Now, since you opened this issue, you may like to comment?

**Comment by fkiraly** [comment]
@tarpas, do you understand why we would need `pyargs`? Sorry for asking a potentially stupid question, would appreciate an explanation like I'm a 6 year old. Here's a PR to a state without it, I think:
https://github.com/alan-turing-institute/sktime/pull/2928
just to see what happens...

**Comment by fkiraly** [comment]
seems to work without `pyargs`, @tarpas
https://github.com/alan-turing-institute/sktime/pull/2928
is this sth you can work with?

**Comment by tarpas** [comment]
Per this: _by @lmmentel in https://github.com/alan-turing-institute/sktime/issues/2389#issuecomment-1087828335_ sktime started using --pyargs because pytest was picking up some files (associated with cython which you don't use anymore anyway) which it shouldn't have been picking. I think --ignore or other means is the right tool to tackle that.

So I don't think there was a good reason and #2928 works for me, please go ahead with it!

**Comment by fkiraly** [comment]
Ah! Well, we killed all cython with fire somewhen around 0.9.0, so there is perhaps no longer a reason to use --pyargs, which would solve everything.

**Comment by tarpas** [comment]
@fkiraly can you point me to a PR where there is suspicion testmon might not pick up changes correctly? Would that be adding a new estimator class? Can you point me to an example? 

I'm not that familiar with your project. My attempts to satisfactorily work with a remote development environment (amd64 linux) failed. So I'll probably have to get a physical Intel/AMD machine, so it might take some time..

**Comment by fkiraly** [comment]
> I'm not that familiar with your project. My attempts to satisfactorily work with a remote development environment (amd64 linux) failed. So I'll probably have to get a physical Intel/AMD machine, so it might take some time

We could try setting one up by screenshare sometime? I know @lmmentel, @freddyaboulton and @mloning all have done that.

>Would that be adding a new estimator class? Can you point me to an example?

Yes, adding a new estimator class that is failing tests would be such an example.
Here is a current PR which adds a new estimator class and fails at the time of writing:

https://github.com/alan-turing-institute/sktime/pull/2349

So, the failure mode for the testmon extension would be failing to pick up the failures.

**Comment by fkiraly** [comment]
@tarpas, pyargs are now gone!

**Comment by lmmentel** [comment]
This are great findings, thanks for sharing @fkiraly! 

Separating testing for core and soft dependencies would be great since it would allow us to have more control of what kinds of tests are applied where. 

I think getting more insights would require profiling the test collection. It might be related by lately I was running a simple `pytest filename.py` but still using our pytest configuration and the startup/collection took way longer that expected ~10 sec for a file that had 1 simple test in it. I didn't investigate further by it might be that test collection is slow. Might be worth doing some checks switching off plugins and timing just the test collection. 

Another thing is sktime's own test automation that should be profiled so that we can see where the time is spent.

**Comment by tarpas** [comment]
I think both profiling the test collection and splitting the whole test suite are quite important steps. Both would be very helpful. The 10 minutes collection time makes any experiments with the test suite dreadful.

**Comment by fkiraly** [comment]
Changing test order reducing expected runtime by 90%: https://github.com/alan-turing-institute/sktime/pull/3092

**Comment by fkiraly** [comment]
Ok, here's another idea: partition matrix design #3137 - reduction of test time to 30-40%

**Comment by fkiraly** [comment]
This wasn't completed, test time is still over 10min.

**Comment by fkiraly** [comment]
Here's a PR that follows an idea based on discussions at EuroPython, FYI @tarpas.
https://github.com/sktime/sktime/pull/5019

The PR introduces logic where we check only estimators from modules that have changed compared to current `main`.

While this works perfectly locally, it seems that the VMs in the CI don't have the `main` branch, so the logic does not work.

This would be solved by ensuring the `main` branch is also present in the VM repos, but not sure how to do that. Any ideas?

**Comment by fkiraly** [comment]
IT WORKS! FYI @tarpas!

**Comment by fkiraly** [comment]
Interestingly, putting test times together from #5019 and current tests shows:

* there's still about 15-20min worth of non-suite tests which do not depend on soft dependencies
* there's about 30min worth of non-suite tests which also depend on soft dependencies
* all suite tests for estimators are worth ca 180 min (slashed to 60min per CI element via the "divide by three" matrix test schema)

**Comment by fkiraly** [comment]
hm, we are not quite there yet.

Full test run without soft dependencies is around 15min, and with is around 25min, so we are close...

Also, what now takes longest is the notebooks - 45min. That also needs to be addressed.

**Comment by yarnabrina** [comment]
This seems to be on the same topic. May be we can leverage it, when it gets ready.

https://about.codecov.io/smart-automated-test-selection/

**Comment by fkiraly** [comment]
hm, @tarpas, have you seen that?

**Comment by yarnabrina** [comment]
> This seems to be on the same topic. May be we can leverage it, when it gets ready.
> 
> https://about.codecov.io/smart-automated-test-selection/

Remembered this after long time. @fkiraly your recent PR's about differential testing is essentially this, right? Only run the tests for modified stuff?

**Comment by fkiraly** [comment]
indeed. The only wrinkle is that codecov now returns far too low coverages.
See your issue https://github.com/sktime/sktime/issues/6035

**Comment by fkiraly** [comment]
I think we can close this issue now, @yarnabrina?

Average test times for "normal" PR are at 5min, if you discount environment install etc.

**Comment by yarnabrina** [comment]
Yes, for normal average PR we should be good. PR's with huge amount of changes will take more time of course, but I feel that's desirable too for sake of coverage/confidence.

We can close this then, once all your differential ones are merged. Is there any left?

**Comment by fkiraly** [comment]
> Is there any left?

Yes, I tag test framework related tests with the `module:tests` tag:

https://github.com/sktime/sktime/pull/6621
https://github.com/sktime/sktime/pull/6620
https://github.com/sktime/sktime/pull/6619
